File: clique/go/bit_vector.go
package clique

type BitVector uint64

func (b *BitVector) Set(i int) {
	*b |= (1 << uint(i))
}

func (b BitVector) Has(i int) bool {
	return b&(1<<uint(i)) != 0
}

func (b BitVector) Intersect(c BitVector) BitVector {
	return b & c
}

func (b BitVector) Slice() []int {
	s := []int{}
	for i := 0; i < 64; i++ {
		if b.Has(i) {
			s = append(s, i)
		}
	}
	return s
}

File: independent_set/rust/bit_vector.rs
pub struct BitVector {
    value: u64,
}

impl BitVector {
    pub fn new() -> Self {
        BitVector { value: 0 }
    }

    pub fn set(&mut self, i: usize) {
        self.value |= 1 << i;
    }

    pub fn has(&self, i: usize) -> bool {
        self.value & 1 << i != 0
    }

    pub fn intersect(&self, other: &Self) -> Self {
        BitVector { value: self.value & other.value }
    }

    pub fn is_empty(&self) -> bool {
        self.value == 0
    }

    pub fn to_vec(&self) -> Vec<usize> {
        let mut vec = vec![];
        for i in 0..64 {
            if self.has(i) {
                vec.push(i);
            }
        }
        vec
    }
}

impl Clone for BitVector {
    fn clone(&self) -> Self {
        BitVector { value: self.value }
    }
}
